### PR TITLE
boards: arm: add support for NUCLEO-L412RB-P

### LINKED ISSUE #33655: Add support for board: Nucleo-L412RB-P
I bought this board to try out the L4 series and noticed L412RB is not supported: https://docs.zephyrproject.org/latest/boards/index.html

I'm thinking about adding this support to Zephyr myself as a challenge/exercise.

* https://www.digikey.com/en/products/detail/stmicroelectronics/NUCLEO-L412RB-P/9656230

-----
Self notes:

* Board porting guide: https://docs.zephyrproject.org/latest/guides/porting/board_porting.html

### SoC Support
1. - [x] Define STM32L412RB  in device tree. (PR: https://github.com/zephyrproject-rtos/zephyr/pull/33660)
2. - [x] The SoC is not fully supported. I need to add something similar for `STM32L412RB` as is done for `STM32L32`: https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_SOC_STM32L432XX.html#cmdoption-arg-CONFIG_SOC_STM32L432XX

----

### Creating files for the board

1. - [x] Write devicetree file: `boards/arm/nucleo_l412rb/nucleo_l412rb.dts` https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#write-your-devicetree
      - - [x] User LED
      - - [x] User buttons
      - - [ ] Add 32.768kHz RTC clock 
2. - [x] Write Kconfig files
3. - [x] Build, test, and fix
    * `west build -b nucleo_l412rb samples/hello_world`
    * `west flash`
4. - [x] Update documentation

----

### Test samples

- [x] Hello World
- [x] Blinky
- [ ] Blinky PWM (compiles and flashes -- but PWM does not work)
- [x] Button
- [ ] Fade LED (compiles and flashes -- but PWM does not work)
- [ ] Threads
- [x] Counter

----
Other commits as examples:
* Commit where support for SoC STM32F410XX was added c99ae8d
* Commit where support for Nucleo L432KC was added: https://github.com/zephyrproject-rtos/zephyr/commit/6256c997034b6188e0dc908d64d73a90b780766e

### PR DESCRIPTION
**Pull request by gudnimg**
This PR adds support for ST Nucleo-64 Development board with STM32L412RB SoC.
Tested samples: hello_world, blinky, button.

Closes https://github.com/zephyrproject-rtos/zephyr/issues/33655

I plan on adding more peripherals for this board after this PR is merged.

Signed-off-by: Guðni Már Gilbert <gudni.m.g@gmail.com>

### CODE REVIEW

**Comment by GunZi200** [comment]
The build fails because I did not use this in the board's device tree:

```
&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		/*
		 * Reserve the final 16 KiB for file system partition
		 */
		storage_partition: partition@3c000 {
			label = "storage";
			reg = <0x0003c000 0x00004000>;
		};
	};
};
```

I did not include it since I do not understand the purpose of this partition or where the register addresses come from (namely `3c000`). I could find no reference for this in the STM32 datasheet/application notes. If anyone could point me in the right direction I will add this back.

I thought maybe this was just SRAM2? For STM32L412RB its 8KB.

**Comment by GunZi200** [comment]
Ok I figured out how to calculate the correct sizes for the 16KiB partition. Will add it in an hour or two.

File: boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+type: mcu
+arch: arm
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
+ram: 40
+flash: 64
+supported:
+  - nvs
```
Comment (erwango) [inline boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.yaml:12]: This is probably the reason why partitions are requested.
This is welcome, but could be removed if not needed

**Review by FRASTM — APPROVED** [review]

**Comment by GunZi200** [comment]
I added the flash partition with correct values.

**Comment by GunZi200** [comment]
Added a missing new line at end of file 😃

**Comment by GunZi200** [comment]
Question regarding the CI... is the CI supposed to skip I2C sensor if the board has I2C peripheral? Just wondering if this is normal or not.

image

**Comment by erwango** [comment]
> Question regarding the CI... is the CI supposed to skip I2C sensor if the board has I2C peripheral? Just wondering if this is normal or not.

This is normal. To "run" these tests, a sensor would be required to be configured on the board, 
as this is not the case on this board, they are skipped.

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Looks good, thank you!

**Pull request merged**